File: 637.二叉树的层平均值.py
#
# @lc app=leetcode.cn id=637 lang=python3
#
# [637] 二叉树的层平均值
#
# https://leetcode-cn.com/problems/average-of-levels-in-binary-tree/description/
#
# algorithms
# Easy (68.84%)
# Likes:    268
# Dislikes: 0
# Total Accepted:    66.2K
# Total Submissions: 96.1K
# Testcase Example:  '[3,9,20,null,null,15,7]'
#
# 给定一个非空二叉树, 返回一个由每层节点平均值组成的数组。
# 
# 
# 
# 示例 1：
# 
# 输入：
# ⁠   3
# ⁠  / \
# ⁠ 9  20
# ⁠   /  \
# ⁠  15   7
# 输出：[3, 14.5, 11]
# 解释：
# 第 0 层的平均值是 3 ,  第1层是 14.5 , 第2层是 11 。因此返回 [3, 14.5, 11] 。
# 
# 
# 
# 
# 提示：
# 
# 
# 节点值的范围在32位有符号整数范围内。
# 
# 
#

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
from collections import deque


class Solution:
    def averageOfLevels(self, root: TreeNode) -> list[float]:
        ret = []
        if not root:
            return ret

        queue = deque([root])
        while queue:
            level_length = len(queue)
            level_ret = []

            for _ in range(level_length):
                curr = queue.popleft()
                level_ret.append(curr.val)

                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)

            ret.append(sum(level_ret) / level_length)

        return ret
# @lc code=end

File: 637.二叉树的层平均值.ts
/*
 * @lc app=leetcode.cn id=637 lang=typescript
 *
 * [637] 二叉树的层平均值
 *
 * https://leetcode.cn/problems/average-of-levels-in-binary-tree/description/
 *
 * algorithms
 * Easy (69.51%)
 * Likes:    344
 * Dislikes: 0
 * Total Accepted:    103.1K
 * Total Submissions: 148.3K
 * Testcase Example:  '[3,9,20,null,null,15,7]'
 *
 * 给定一个非空二叉树的根节点 root , 以数组的形式返回每一层节点的平均值。与实际答案相差 10^-5 以内的答案可以被接受。
 *
 *
 *
 * 示例 1：
 *
 *
 *
 *
 * 输入：root = [3,9,20,null,null,15,7]
 * 输出：[3.00000,14.50000,11.00000]
 * 解释：第 0 层的平均值为 3,第 1 层的平均值为 14.5,第 2 层的平均值为 11 。
 * 因此返回 [3, 14.5, 11] 。
 *
 *
 * 示例 2:
 *
 *
 *
 *
 * 输入：root = [3,9,20,15,7]
 * 输出：[3.00000,14.50000,11.00000]
 *
 *
 *
 *
 * 提示：
 *
 *
 *
 *
 * 树中节点数量在 [1, 10^4] 范围内
 * -2^31 <= Node.val <= 2^31 - 1
 *
 *
 */

import { TreeNode } from './commons/Tree'

export
// @lc code=start
function averageOfLevels(root: TreeNode | null): number[] {
  const result: number[] = []
  const queue = [root]

  while (queue.length) {
    let levelSum = 0
    const levelLength = queue.length

    for (let i = 0; i < levelLength; i++) {
      const node = queue.shift()!

      levelSum += node.val

      if (node.left) {
        queue.push(node.left)
      }
      if (node.right) {
        queue.push(node.right)
      }
    }

    result.push(levelSum / levelLength)
  }

  return result
}
// @lc code=end
